src/test/scala/com/microsoft/hyperspace/index/rules/JoinIndexRuleV2Test.scala

```diff
@@ -0,0 +1,247 @@
+
+  test("Join rule does not update plan for Broadcast Hash Join compatible joins") {
+    withSQLConf("spark.hyperspace.rule.joinV2.enabled" -> "true") {
+      val joinCondition = EqualTo(t1c1, t3c1)
+      val originalPlan =
+        Join(t1ProjectNode, t3ProjectNode, JoinType("inner"), Some(joinCondition))
+      val updatedPlan = JoinIndexRuleV2(originalPlan)
+      assert(updatedPlan.equals(originalPlan))
+    }
+  }
+
+  test("Join rule works if indexes exist and configs are set correctly") {
+    withSQLConf("spark.hyperspace.rule.joinV2.enabled" -> "true") {
+      val joinCondition = EqualTo(t1c1, t2c1)
+      val originalPlan =
+        Join(t1ProjectNode, t2ProjectNode, JoinType("inner"), Some(joinCondition))
+      val updatedPlan = JoinIndexRuleV2(originalPlan)
+      assert(!updatedPlan.equals(originalPlan))
+
+      val indexPaths = getIndexDataFilesPaths("t1i1") ++ getIndexDataFilesPaths("t2i1")
+      verifyUpdatedIndex(originalPlan, updatedPlan, indexPaths)
+    }
+  }
+
+  test("Join rule updates sort merge join part of a plan with both smj and bhj.") {
+    withSQLConf("spark.hyperspace.rule.joinV2.enabled" -> "true") {
+      val bhjCondition = EqualTo(t2c1, t3c1)
+      val bhjJoin = Join(t2ProjectNode, t3ProjectNode, JoinType("inner"), Some(bhjCondition))
+
+      val smjCondition = EqualTo(t1c1, t2c1)
+      val originalPlan = Join(t1ProjectNode, bhjJoin, JoinType("inner"), Some(smjCondition))
+
+      val updatedPlan = JoinIndexRuleV2(originalPlan)
+      assert(!updatedPlan.equals(originalPlan))
+
+      val indexPaths =
+        getIndexDataFilesPaths("t1i1") ++ getIndexDataFilesPaths("t2i1")
+      verifyUpdatedIndex(originalPlan, updatedPlan, indexPaths ++ Seq(new Path("t3")))
+    }
+  }
+
+  private def verifyUpdatedIndex(
+      originalPlan: Join,
+      updatedPlan: LogicalPlan,
+      indexPaths: Seq[Path]): Unit = {
+    assert(treeStructureEquality(originalPlan, updatedPlan))
+    assert(basePaths(updatedPlan) == indexPaths)
+  }
+
+  /** Method to check if tree structures of two logical plans are the same. */
+  private def treeStructureEquality(plan1: LogicalPlan, plan2: LogicalPlan): Boolean = {
+    val originalNodeCount = plan1.treeString.split("\n").length
+    val updatedNodeCount = plan2.treeString.split("\n").length
+
+    if (originalNodeCount == updatedNodeCount) {
+      (0 until originalNodeCount).forall { i =>
+        plan1(i) match {
+          // For LogicalRelation, we just check if the updated also has LogicalRelation. If the
+          // updated plan uses index, the root paths will be different here.
+          case _: LogicalRelation => plan2(i).isInstanceOf[LogicalRelation]
+
+          // For other node types, we compare exact matching between original and updated plans.
+          case node => node.simpleString.equals(plan2(i).simpleString)
+        }
+      }
+    } else {
+      false
+    }
+  }
+
+  /** Returns tuple of left and right base relation paths for a logical plan. */
```

Comment:
Than use Tuple as a return type?

Reply:
thanks @imback82 , improved the comment since the method is more generic than tuples.